In [11]:
import numpy as np
import scipy.stats as stats
import pandas as pd
from statsmodels.stats.multicomp import pairwise_tukeyhsd

# Домашняя работа по уроку 10. Дисперсионный анализ

## Задача 1

Провести дисперсионный анализ для определения того, 

есть ли различия среднего роста среди взрослых футболистов, хоккеистов и штангистов.

Даны значения роста в трех группах случайно выбранных спортсменов:

Футболисты: 173, 175, 180, 178, 177, 185, 183, 182.

Хоккеисты: 177, 179, 180, 188, 177, 172, 171, 184, 180.

Штангисты: 172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170.

### Решение

In [2]:
gr1 = np.array([173, 175, 180, 178, 177, 185, 183, 182])
gr2 = np.array([177, 179, 180, 188, 177, 172, 171, 184, 180])
gr3 = np.array([172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170])

Проведем разведочный анализ (EDA)

In [4]:
# Проверка на нормальность распределения (тест Шапиро-Уилка)
stats.shapiro(gr1)

ShapiroResult(statistic=0.9775082468986511, pvalue=0.9495404362678528)

pvalue > $\alpha$ = 0.05, =>, H<sub>0</sub> о нормальности распределения верна.

Распределение для среднего роста футболистов является нормальным.

In [5]:
# Проверка на нормальность распределения (тест Шапиро-Уилка)
stats.shapiro(gr2)

ShapiroResult(statistic=0.9579196572303772, pvalue=0.7763139009475708)

pvalue > $\alpha$ = 0.05, =>, H<sub>0</sub> о нормальности распределения верна.

Распределение для среднего роста хоккеистов является нормальным.

In [6]:
# Проверка на нормальность распределения (тест Шапиро-Уилка)
stats.shapiro(gr3)

ShapiroResult(statistic=0.9386808276176453, pvalue=0.5051165223121643)

pvalue > $\alpha$ = 0.05, =>, H<sub>0</sub> о нормальности распределения верна.

Распределение для среднего роста штангистов является нормальным.

Таким образом, условие нормальности распределения выполняется.

Выборки являются независимыми.

In [9]:
# Проверим однородность дисперсий Барлетт тестом
stats.bartlett(gr1, gr2, gr3)

BartlettResult(statistic=0.4640521043406442, pvalue=0.7929254656083131)

Получили pvalue = 0.79 > $\alpha$ = 0.05, =>, H<sub>0</sub> об однородности дисперсий верна.

Таким образом, уловия применимости дисперсионного анализа выполняются.

In [10]:
# однофакторный дисперсионный анализ
stats.f_oneway(gr1, gr2, gr3)

F_onewayResult(statistic=5.500053450812596, pvalue=0.010482206918698693)

В результате однофакторного дисперсионного анализа получили

pvalue = 0.01 < $\alpha$ = 0.05, =>,

H<sub>0</sub> об отсутствии различий между группами отвергается,

есть статистически значимые различия.

Рассмотрим, между какими именно выборками имеются различия с помощью тета Тьюки:

In [12]:
df = pd.DataFrame({"score" : [173, 175, 180, 178, 177, 185, 183, 182,
                              177, 179, 180, 188, 177, 172, 171, 184, 180,
                              172, 173, 169, 177, 166, 180, 178, 177, 172, 166, 170],
                   "group" : np.repeat(["gr1", "gr2", "gr3"], repeats=[8, 9, 11])})
df

score group
0     173   gr1
1     175   gr1
2     180   gr1
3     178   gr1
4     177   gr1
5     185   gr1
6     183   gr1
7     182   gr1
8     177   gr2
9     179   gr2
10    180   gr2
11    188   gr2
12    177   gr2
13    172   gr2
14    171   gr2
15    184   gr2
16    180   gr2
17    172   gr3
18    173   gr3
19    169   gr3
20    177   gr3
21    166   gr3
22    180   gr3
23    178   gr3
24    177   gr3
25    172   gr3
26    166   gr3
27    170   gr3

In [15]:
tukey = pairwise_tukeyhsd(df["score"], df["group"], alpha = 0.05)
print(tukey)

 Multiple Comparison of Means - Tukey HSD, FWER=0.05 
group1 group2 meandiff p-adj   lower    upper  reject
-----------------------------------------------------
   gr1    gr2  -0.4583  0.979  -6.2732  5.3566  False
   gr1    gr3  -6.3977 0.0219 -11.9583 -0.8372   True
   gr2    gr3  -5.9394 0.0284 -11.3181 -0.5607   True
-----------------------------------------------------


Таким образом, видим, что средний рост футболистов и хоккеистов не имеет

статистически значимых различий (pvalue = 0.98 > $\alpha$). А средний рост между футболистами и штангистами(pvalue = 0.0219 > $\alpha$),

хоккеистами и штангистами (pvalue = 0.0284 > $\alpha$) имеет статистически значимые различия.